### ISSUE TITLE
Only alphanumeric character set ranges are detected as ranges

### ISSUE THREAD

**Issue by janosch-x** [ammar/regexp_parser#29]
```ruby
S = Regexp::Scanner
S.scan /[1-9]/ # => [..., [:set, :range, "1-9", 1, 4], ...]
S.scan /[a-z]/ # => [..., [:set, :range, "a-z", 1, 4], ...]
S.scan /[!-%]/ # => [..., [:set, :member, "!", 1, 2], [:set, :member, "-", 2, 3], [:set, :member, "%", 3, 4], ...]
S.scan /[ä-ü]/ # => [..., [:set, :member, "\xC3\xA4", 1, 3], [:set, :member, "-", 3, 4], [:set, :member, "\xC3\xBC", 4, 6], ...]
```

I think we could detect everything as a range that follows this pattern:

```
anything_but_a_bracket . "-" . anything_but_a_bracket
```

Of course, "anything" can be a unicode escape or even a codepoint list. Funny enough this works:

```ruby
/[\u{41 42}-\u{55 56}]/ =~ "\u{44}" # => 0
```

But besides of that are there any other pitfalls I didn’t think of?

**Comment by ammar** [comment]
That change makes sense. All of those cases should identified as ranges by the scanner, leaving it up to the lexer or parser to handle the semantics of the range, and possibly detect invalid ones.

**Comment by janosch-x** [comment]
I’ve started playing with scanner.rl to fix this issue, but come across a few other issues along the way. 

Similar to the point made about sets in #28, marking stuff as a `range` causes a loss of token information. If `0-9`, `\a-\e` and `\u{600}-\u{6FF}` are all emitted simply as `range`, library users will have to scan the token data themselves if they want to act on specific token types.

Encoding both the "rangeness" and token type at index 1 sounds like an option, and is already done for `:range_hex`, but it does not seem like a sustainable solution. It would require lots of new token types to cover all cases, and some cases are ambiguous. E.g. is `\x42-\u{6FF}` a `range_hex`, a `range_codepoint`, or something else altogether?

So I’m a little unsure whether a bigger refactoring might make sense.

A first step could still be to simply detect all ranges as `range`. However, even that first step is bigger than I initially thought, with chars like `^` or `\` potentially acting as the start or end of a range. So I guess I’ll wait and see what conclusion you reach about #28 to avoid duplicate work on these parts.

**Comment by janosch-x** [comment]
fixed by #55
